src/Jackalope/Version/VersionHandler.php

```diff
@@ -245,5 +247,172 @@ private function createFrozenNode(NodeInterface $versionNode, NodeInterface $nod
 
             $frozenNode->setProperty($propertyName, $property->getValue());
         }
+
+        foreach ($node->getNodes() as $childNode) {
+            /** @var NodeInterface $childNode */
+            // TODO also use the commented lines when getDefinition is implemented
+            //$onParentVersion = $childNode->getDefinition()->getOnParentVersion();
+
+            //if ($onParentVersion == OnParentVersionAction::COPY) {
+            // can't use the workspace copy command, because the parent node already has to exist
+            // TODO distinguish between COPY and VERSION onParentVersion
+            $this->copyIntoNode($childNode, $frozenNode);
+            //}
+        }
+    }
+
+    public function restoreItem($removeExisting, $versionPath, $path)
+    {
+        $node = $this->objectManager->getNodeByPath($path);
+
+        if ($node->isModified()) {
+            throw new InvalidItemStateException(
+                sprintf(
+                    'Node "%s" contains unsaved changes',
+                    $path
+                )
+            );
+        }
+
+        $versionNode = $this->objectManager->getNodeByPath($versionPath, 'Version\Version');
+        $frozenNode = $versionNode->getNode('jcr:frozenNode');
+
+        if ($frozenNode->getPropertyValue('jcr:frozenUuid') != $node->getPropertyValue('jcr:uuid')) {
+            throw new RepositoryException(sprintf(
+                'The frozen node uuid "%s" does not match the actual node uuid "%s". This should never happen.',
+                $frozenNode->getPropertyValue('jcr:frozenUuid'),
+                $node->getPropertyValue('jcr:uuid'))
+            );
+        }
+
+        // TODO reset primary type once the primary type can be changed
+        // also see (https://github.com/jackalope/jackalope/issues/247)
```

Comment:
imho we can ignore this for now. as long as the primary type can't be changed, there is also no need to restore. lets keep the comment here though, it would make sense to eventually support this feature.